File: android/app/src/main/java/local/gestorweb/android/LicenseClient.java
package local.gestorweb.android;

import android.content.Context;
import org.json.JSONArray;
import org.json.JSONObject;
import java.io.BufferedReader;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.OutputStream;
import java.net.HttpURLConnection;
import java.net.URL;
import java.nio.charset.StandardCharsets;
import java.security.KeyFactory;
import java.security.MessageDigest;
import java.security.PublicKey;
import java.security.Signature;
import java.security.spec.X509EncodedKeySpec;
import java.util.Base64;

final class LicenseClient {
    private static final String PREFIX = "GW-LIC-V1";
    private static final String APP_ID = "gestor-web";

    static final class Result {
        final boolean active;
        final String reason;
        final long expiresAt;
        Result(boolean active, String reason, long expiresAt) {
            this.active = active; this.reason = reason; this.expiresAt = expiresAt;
        }
    }

    private static final class ParsedLicense {
        final String normalized;
        final String payloadSegment;
        final byte[] signature;
        final JSONObject payload;
        ParsedLicense(String normalized, String payloadSegment, byte[] signature, JSONObject payload) {
            this.normalized = normalized;
            this.payloadSegment = payloadSegment;
            this.signature = signature;
            this.payload = payload;
        }
    }

    static Result verify(Context context, String licenseText, String deviceId) {
        try {
            ParsedLicense parsed = parseLicense(licenseText);
            if (!verifySignature(context, parsed)) return new Result(false, "firma criptografica invalida", 0);

            String app = parsed.payload.optString("app", APP_ID);
            if (!APP_ID.equals(app)) return new Result(false, "licencia emitida para otra app", 0);

            String hwid = parsed.payload.optString("hwid", "");
            if (!hwid.isEmpty() && !hwid.equals(deviceId)) return new Result(false, "licencia emitida para otro dispositivo", 0);

            long expiresAt = expiresAt(parsed.payload);
            if (expiresAt > 0 && System.currentTimeMillis() > expiresAt) return new Result(false, "licencia expirada", expiresAt);

            JSONObject db = parsed.payload.optJSONObject("license_db");
            if (db == null || !"supabase".equalsIgnoreCase(db.optString("provider", ""))) {
                return new Result(false, "licencia sin configuracion Supabase", expiresAt);
            }

            String supabaseUrl = db.optString("url", "").replaceAll("/+$", "");
            String anonKey = db.optString("anon_key", "");
            if (supabaseUrl.isEmpty() || anonKey.isEmpty()) return new Result(false, "Supabase URL o anon key no configurados", expiresAt);

            return verifySupabase(parsed, deviceId, supabaseUrl, anonKey, expiresAt);
        } catch (Exception error) {
            return new Result(false, "No se pudo validar licencia: " + error.getMessage(), 0);
        }
    }

    private static ParsedLicense parseLicense(String text) throws Exception {
        String normalized = normalizeLicenseText(text);
        if (!normalized.startsWith(PREFIX)) throw new IllegalArgumentException("prefijo invalido");

        String body = normalized.substring(PREFIX.length()).trim();
        if (body.startsWith(":") || body.startsWith(".")) body = body.substring(1).trim();

        String payloadSegment = "";
        String signatureSegment = "";
        String[] lines = body.split("\\n+");
        if (lines.length >= 2) {
            payloadSegment = lines[0].trim();
            signatureSegment = lines[1].trim();
        } else {
            String[] parts = body.split("\\.");
            if (parts.length >= 2) {
                payloadSegment = parts[0].trim();
                signatureSegment = parts[1].trim();
            }
        }
        if (payloadSegment.isEmpty() || signatureSegment.isEmpty()) throw new IllegalArgumentException("formato invalido");

        byte[] payloadBytes = Base64.getUrlDecoder().decode(payloadSegment);
        byte[] signature = Base64.getUrlDecoder().decode(signatureSegment);
        JSONObject payload = new JSONObject(new String(payloadBytes, StandardCharsets.UTF_8));
        return new ParsedLicense(normalized, payloadSegment, signature, payload);
    }

    private static String normalizeLicenseText(String text) {
        return String.valueOf(text == null ? "" : text).replace("\r\n", "\n").replace("\r", "\n").trim();
    }

    private static boolean verifySignature(Context context, ParsedLicense parsed) throws Exception {
        Signature verifier = Signature.getInstance("SHA256withRSA");
        verifier.initVerify(loadPublicKey(context));
        verifier.update(parsed.payloadSegment.getBytes(StandardCharsets.UTF_8));
        return verifier.verify(parsed.signature);
    }

    private static PublicKey loadPublicKey(Context context) throws Exception {
        StringBuilder pem = new StringBuilder();
        try (InputStream input = context.getAssets().open("license-public-key.pem");
             BufferedReader reader = new BufferedReader(new InputStreamReader(input, StandardCharsets.UTF_8))) {
            String line;
            while ((line = reader.readLine()) != null) pem.append(line).append('\n');
        }
        String clean = pem.toString()
            .replace("-----BEGIN PUBLIC KEY-----", "")
            .replace("-----END PUBLIC KEY-----", "")
            .replaceAll("\\s", "");
        if (clean.isEmpty() || clean.contains("REPLACE")) throw new IllegalStateException("clave publica no configurada en assets/license-public-key.pem");
        byte[] encoded = Base64.getDecoder().decode(clean);
        return KeyFactory.getInstance("RSA").generatePublic(new X509EncodedKeySpec(encoded));
    }

    private static long expiresAt(JSONObject payload) {
        if (!payload.has("expires_at") || payload.isNull("expires_at")) return 0;
        return payload.optLong("expires_at", 0);
    }

    private static Result verifySupabase(ParsedLicense parsed, String deviceId, String supabaseUrl, String anonKey, long localExpiresAt) {
        HttpURLConnection connection = null;
        try {
            URL url = new URL(supabaseUrl + "/rest/v1/rpc/verify_license_public");
            connection = (HttpURLConnection) url.openConnection();
            connection.setRequestMethod("POST");
            connection.setConnectTimeout(8000);
            connection.setReadTimeout(8000);
            connection.setDoOutput(true);
            connection.setRequestProperty("content-type", "application/json");
            connection.setRequestProperty("apikey", anonKey);
            connection.setRequestProperty("Authorization", "Bearer " + anonKey);

            JSONObject request = new JSONObject();
            request.put("p_id", parsed.payload.optString("id", ""));
            request.put("p_hwid", deviceId);
            request.put("p_license_hash", sha256(parsed.normalized));
            try (OutputStream output = connection.getOutputStream()) {
                output.write(request.toString().getBytes(StandardCharsets.UTF_8));
            }

            int status = connection.getResponseCode();
            InputStream stream = status >= 200 && status < 300 ? connection.getInputStream() : connection.getErrorStream();
            String body = readBody(stream);
            if (status < 200 || status >= 300) return new Result(false, "Supabase HTTP " + status + ": " + body, localExpiresAt);

            JSONObject response;
            if (body.trim().startsWith("[")) response = new JSONArray(body).optJSONObject(0);
            else response = new JSONObject(body);
            if (response == null) return new Result(false, "respuesta Supabase vacia", localExpiresAt);

            boolean active = response.optBoolean("active", false);
            String reason = response.optString("reason", active ? "ok" : "licencia no activa");
            long expiresAt = response.has("expires_at") && !response.isNull("expires_at") ? response.optLong("expires_at", localExpiresAt) : localExpiresAt;
            return new Result(active, reason, expiresAt);
        } catch (Exception error) {
            return new Result(false, "No se pudo validar online: " + error.getMessage(), localExpiresAt);
        } finally {
            if (connection != null) connection.disconnect();
        }
    }

    private static String readBody(InputStream stream) throws Exception {
        if (stream == null) return "";
        StringBuilder body = new StringBuilder();
        try (BufferedReader reader = new BufferedReader(new InputStreamReader(stream, StandardCharsets.UTF_8))) {
            String line;
            while ((line = reader.readLine()) != null) body.append(line);
        }
        return body.toString();
    }

    private static String sha256(String text) throws Exception {
        byte[] digest = MessageDigest.getInstance("SHA-256").digest(text.getBytes(StandardCharsets.UTF_8));
        StringBuilder hex = new StringBuilder();
        for (byte value : digest) hex.append(String.format("%02x", value));
        return hex.toString();
    }
}

File: android/app/src/main/java/local/gestorweb/android/LicenseClient.kt
package local.gestorweb.android

import android.content.Context
import org.json.JSONArray
import org.json.JSONObject
import java.io.BufferedReader
import java.io.InputStream
import java.io.InputStreamReader
import java.io.OutputStream
import java.net.HttpURLConnection
import java.net.URL
import java.nio.charset.StandardCharsets
import java.security.KeyFactory
import java.security.MessageDigest
import java.security.PublicKey
import java.security.Signature
import java.security.spec.X509EncodedKeySpec
import java.util.Base64

internal object LicenseClient {

    private const val PREFIX = "GW-LIC-V1"
    private const val APP_ID = "gestor-web"

    data class Result(val active: Boolean, val reason: String, val expiresAt: Long)

    private data class ParsedLicense(
        val normalized: String,
        val payloadSegment: String,
        val signature: ByteArray,
        val payload: JSONObject
    )

    // ── Public entry point ───────────────────────────────────────────────────

    fun verify(context: Context, licenseText: String, deviceId: String): Result = try {
        val parsed = parseLicense(licenseText)
        if (!verifySignature(context, parsed)) return Result(false, "firma criptografica invalida", 0)

        val app = parsed.payload.optString("app", APP_ID)
        if (app != APP_ID) return Result(false, "licencia emitida para otra app", 0)

        val hwid = parsed.payload.optString("hwid", "")
        if (hwid.isNotEmpty() && hwid != deviceId) return Result(false, "licencia emitida para otro dispositivo", 0)

        val expiresAt = expiresAt(parsed.payload)
        if (expiresAt > 0 && System.currentTimeMillis() > expiresAt) return Result(false, "licencia expirada", expiresAt)

        val db = parsed.payload.optJSONObject("license_db")
            ?: return Result(false, "licencia sin configuracion Supabase", expiresAt)
        if (!db.optString("provider", "").equals("supabase", ignoreCase = true))
            return Result(false, "licencia sin configuracion Supabase", expiresAt)

        val supabaseUrl = db.optString("url", "").trimEnd('/')
        val anonKey    = db.optString("anon_key", "")
        if (supabaseUrl.isEmpty() || anonKey.isEmpty())
            return Result(false, "Supabase URL o anon key no configurados", expiresAt)

        verifySupabase(parsed, deviceId, supabaseUrl, anonKey, expiresAt)
    } catch (e: Exception) {
        Result(false, "No se pudo validar licencia: ${e.message}", 0)
    }

    // ── Parsing ───────────────────────────────────────────────────────────────

    private fun parseLicense(text: String): ParsedLicense {
        val normalized = normalize(text)
        if (!normalized.startsWith(PREFIX)) throw IllegalArgumentException("prefijo invalido")

        var body = normalized.substring(PREFIX.length).trim()
        if (body.startsWith(":") || body.startsWith(".")) body = body.substring(1).trim()

        val lines = body.split(Regex("\\n+"))
        val (payloadSegment, signatureSegment) = if (lines.size >= 2) {
            lines[0].trim() to lines[1].trim()
        } else {
            val parts = body.split(".")
            if (parts.size >= 2) parts[0].trim() to parts[1].trim()
            else throw IllegalArgumentException("formato invalido")
        }
        if (payloadSegment.isEmpty() || signatureSegment.isEmpty()) throw IllegalArgumentException("formato invalido")

        val payloadBytes = Base64.getUrlDecoder().decode(payloadSegment)
        val signature    = Base64.getUrlDecoder().decode(signatureSegment)
        val payload      = JSONObject(String(payloadBytes, StandardCharsets.UTF_8))
        return ParsedLicense(normalized, payloadSegment, signature, payload)
    }

    private fun normalize(text: String?): String =
        (text ?: "").replace("\r\n", "\n").replace("\r", "\n").trim()

    // ── Signature ─────────────────────────────────────────────────────────────

    private fun verifySignature(context: Context, parsed: ParsedLicense): Boolean {
        val verifier = Signature.getInstance("SHA256withRSA")
        verifier.initVerify(loadPublicKey(context))
        verifier.update(parsed.payloadSegment.toByteArray(StandardCharsets.UTF_8))
        return verifier.verify(parsed.signature)
    }

    private fun loadPublicKey(context: Context): PublicKey {
        val pem = context.assets.open("license-public-key.pem").bufferedReader().use { it.readText() }
        val clean = pem
            .replace("-----BEGIN PUBLIC KEY-----", "")
            .replace("-----END PUBLIC KEY-----", "")
            .replace(Regex("\\s"), "")
        if (clean.isEmpty() || clean.contains("REPLACE"))
            throw IllegalStateException("clave publica no configurada en assets/license-public-key.pem")
        val encoded = Base64.getDecoder().decode(clean)
        return KeyFactory.getInstance("RSA").generatePublic(X509EncodedKeySpec(encoded))
    }

    // ── Expiry ────────────────────────────────────────────────────────────────

    private fun expiresAt(payload: JSONObject): Long =
        if (!payload.has("expires_at") || payload.isNull("expires_at")) 0L
        else payload.optLong("expires_at", 0L)

    // ── Supabase online check ─────────────────────────────────────────────────

    private fun verifySupabase(
        parsed: ParsedLicense,
        deviceId: String,
        supabaseUrl: String,
        anonKey: String,
        localExpiresAt: Long
    ): Result {
        var connection: HttpURLConnection? = null
        return try {
            val url = URL("$supabaseUrl/rest/v1/rpc/verify_license_public")
            connection = (url.openConnection() as HttpURLConnection).apply {
                requestMethod = "POST"
                connectTimeout = 8_000
                readTimeout    = 8_000
                doOutput = true
                setRequestProperty("content-type", "application/json")
                setRequestProperty("apikey", anonKey)
                setRequestProperty("Authorization", "Bearer $anonKey")
            }
            val body = JSONObject().apply {
                put("p_id",           parsed.payload.optString("id", ""))
                put("p_hwid",         deviceId)
                put("p_license_hash", sha256(parsed.normalized))
            }.toString().toByteArray(StandardCharsets.UTF_8)
            connection.outputStream.use { it.write(body) }

            val status = connection.responseCode
            val stream = if (status in 200..299) connection.inputStream else connection.errorStream
            val responseBody = readBody(stream)
            if (status !in 200..299) return Result(false, "Supabase HTTP $status: $responseBody", localExpiresAt)

            val response: JSONObject = if (responseBody.trim().startsWith("["))
                JSONArray(responseBody).optJSONObject(0)
                    ?: return Result(false, "respuesta Supabase vacia", localExpiresAt)
            else JSONObject(responseBody)

            val active    = response.optBoolean("active", false)
            val reason    = response.optString("reason", if (active) "ok" else "licencia no activa")
            val expiresAt = if (response.has("expires_at") && !response.isNull("expires_at"))
                response.optLong("expires_at", localExpiresAt) else localExpiresAt
            Result(active, reason, expiresAt)
        } catch (e: Exception) {
            Result(false, "No se pudo validar online: ${e.message}", localExpiresAt)
        } finally {
            connection?.disconnect()
        }
    }

    // ── Utilities ─────────────────────────────────────────────────────────────

    private fun readBody(stream: InputStream?): String {
        stream ?: return ""
        return BufferedReader(InputStreamReader(stream, StandardCharsets.UTF_8)).use { it.readText() }
    }

    private fun sha256(text: String): String {
        val digest = MessageDigest.getInstance("SHA-256").digest(text.toByteArray(StandardCharsets.UTF_8))
        return digest.joinToString("") { "%02x".format(it) }
    }
}
